### PR TITLE
HLD solution for Promotion Counting(Platinum)

### PR DESCRIPTION
**Pull request by Vaishnav-K787**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

File: solutions/gold/usaco-696.mdx [hunk 1]
```diff
@@ -306,10 +307,117 @@ int main() {
...
+using namespace std;
+
+class segTree {
+  private:
+	vector<int> segtree;
 
-Sort nodes from largest to smallest value, do path updates from itself the root.
+  public:
+	segTree(int N) {
+		if (__builtin_popcount(N) > 1) N = 1 << (32 - __builtin_clz(N));
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:329]: why would you need to do a `__builtin_popcount` check? idk seems a bit odd to me

File: solutions/gold/usaco-696.mdx [hunk 2]
```diff
@@ -306,10 +307,117 @@ int main() {
...
+Sort the nodes from largest to smallest value, do path updates from each node to the root.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+class segTree {
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:323]: consider using the iterative sum segtree in the [purs module](https://usaco.guide/gold/PURS?lang=cpp), as I think that impl is cleaner and also probably more familiar

File: solutions/gold/usaco-696.mdx [hunk 3]
```diff
@@ -306,10 +307,117 @@ int main() {
...
+		if (heavy_child == -1 || sizes[heavy_child] < sizes[i]) heavy_child = i;
+	}
+	if (heavy_child != -1) {
+		swap(heavy_paths[start], heavy_paths[heavy_child]);
+		heavy_paths[start].push_back(heavy_child);
+	}
+}
+
+int main() {
+	ios_base::sync_with_stdio(false);
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:363]: bit of a nitpick but typically guide code doesn't use fast io unless needed

File: solutions/gold/usaco-696.mdx [hunk 4]
```diff
@@ -306,10 +307,117 @@ int main() {
...
+	cin.tie(nullptr);
+	freopen("promote.in", "r", stdin);
+	freopen("promote.out", "w", stdout);
+	int N;
+	cin >> N;
+	vector<pair<int, int>> cow_proficiency(N, pair<int, int>());
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:369]: you realize vectors autofill right

File: solutions/gold/usaco-696.mdx [hunk 5]
```diff
@@ -306,10 +307,117 @@ int main() {
...
+	for (int i = 0; i < N; ++i) {
+		cin >> cow_proficiency[i].first;
+		cow_proficiency[i].second = i;
+	}
+	vector<vector<int>> edges(N, vector<int>()), heavy_paths(N, vector<int>());
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:374]: idrc if u change this, but a neat C++ thing is that you can declare 2d vectors like this:

`vector edges(n, vector<int>());`

File: solutions/gold/usaco-696.mdx [hunk 6]
```diff
@@ -306,10 +307,117 @@ int main() {
...
+	for (int i = 0; i < N; ++i) {
+		reverse(heavy_paths[i].begin(), heavy_paths[i].end());
+		for (int j = 0; j < (int)heavy_paths[i].size(); ++j) {
+			heavy_index[heavy_paths[i][j]] = i;
+			path_index[heavy_paths[i][j]] = j;
+		}
+		segtrees[i] = segTree((int)heavy_paths[i].size());
+	}
+
+	sort(cow_proficiency.begin(), cow_proficiency.end(), greater<pair<int, int>>());
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:395]: `sort(cow_proficiency.begin(), cow_proficiency.end(), greater())` also works btw

File: solutions/gold/usaco-696.mdx [hunk 7]
```diff
@@ -306,10 +307,117 @@ int main() {
...
+				}
+				p.second = parents[heavy_paths[heavy_index[p.second]][0]];
+			} else {
+				++cur_ans[p.second];
+				p.second = parents[p.second];
+			}
+		}
+	}
+	for (int i = 0; i < N; ++i) cout << ans[i] << '\n';
+	return 0;
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:418]: this isn't needed

File: solutions/gold/usaco-696.mdx [hunk 8]
```diff
@@ -306,10 +307,117 @@ int main() {
...
+					                                       -1);
+				}
+				p.second = parents[heavy_paths[heavy_index[p.second]][0]];
+			} else {
+				++cur_ans[p.second];
+				p.second = parents[p.second];
+			}
+		}
+	}
+	for (int i = 0; i < N; ++i) cout << ans[i] << '\n';
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:417]: braces

File: solutions/gold/usaco-696.mdx (hunk 6 above)
Comment (Vaishnav-K787) [reply solutions/gold/usaco-696.mdx:398]: it worked locally but threw a compile error on usaco

Comment (TheGamingMousse) [reply solutions/gold/usaco-696.mdx:398]: most guide code is written to be submitted in C++ 17 I'm pretty sure

cuz like, with C++ there's literally no disadvantage with going with a higher version (I'm pretty sure)

if ur curious why, it's cuz of class template argument deduction

File: solutions/gold/usaco-696.mdx [hunk 9]
```diff
@@ -306,10 +307,122 @@ int main() {
 </CPPSection>
 </LanguageSection>
 
-<IncompleteSection>
+## Solution 3: HLD
 
-Add Centroid Decomp/HLD Solution
+**Time Complexity:** $\mathcal{O}(N\log ^2N)$
+
+Sort the nodes from largest to smallest value, do path updates from each node to the root.
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:314]: i get what you mean but like, maybe make the wording a bit more specific 

like: "Sort the nodes in descending order of proficiency. When adding in each node, we do a path update from the root to this node." (except be specific with what type of path update we do)

File: solutions/gold/usaco-696.mdx [hunk 10]
```diff
@@ -306,10 +307,122 @@ int main() {
...
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+template <class T> class SumSegmentTree {
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:323]: put this in a [codesnip](https://usaco.guide/general/working-mdx#code-blocks-and-code-snippets)

File: solutions/gold/usaco-696.mdx [hunk 11]
```diff
@@ -306,10 +307,122 @@ int main() {
...
+		T sum = DEFAULT;
+		for (start += len, end += len; start < end; start /= 2, end /= 2) {
+			if (start % 2 == 1) { sum += segtree[start++]; }
+			if (end % 2 == 1) { sum += segtree[--end]; }
+		}
+		return sum;
+	}
+};
+
+void dfs_hld(vector<vector<int>> &edges, vector<vector<int>> &heavy_paths,
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:351]: ok i personally find this to be kind of ugly, ngl

IMO there's two ways to clean it up
1. make these global arrays (sans might make u change it tho lol)
2. lambda dfs

I would say don't change it yet, but be aware that you might be asked to later.

File: solutions/gold/usaco-696.mdx [hunk 12]
```diff
@@ -306,10 +307,122 @@ int main() {
...
+             vector<int> &depth, vector<int> &sizes, int start) {
+	int heavy_child = -1;
+	++sizes[start];
+	for (int i : edges[start]) {
+		depth[i] = depth[start] + 1;
+		dfs_hld(edges, heavy_paths, depth, sizes, i);
+		sizes[start] += sizes[i];
+		if (heavy_child == -1 || sizes[heavy_child] < sizes[i]) heavy_child = i;
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:359]: just in general put braces around blocks of code even if it's one statement

File: solutions/gold/usaco-696.mdx [hunk 13]
```diff
@@ -306,10 +307,122 @@ int main() {
...
+		swap(heavy_paths[start], heavy_paths[heavy_child]);
+		heavy_paths[start].push_back(heavy_child);
+	}
+}
 
-Sort nodes from largest to smallest value, do path updates from itself the root.
+int main() {
+	freopen("promote.in", "r", stdin);
+	freopen("promote.out", "w", stdout);
+	int N, current_value;
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:370]: very, very, very minor but preferably use `n` instead of `N`

File: solutions/gold/usaco-696.mdx [hunk 14]
```diff
@@ -306,10 +307,122 @@ int main() {
...
+		reverse(heavy_paths[i].begin(), heavy_paths[i].end());
+		for (int j = 0; j < (int)heavy_paths[i].size(); ++j) {
+			heavy_index[heavy_paths[i][j]] = i;
+			path_index[heavy_paths[i][j]] = j;
+		}
+		segtrees[i] = SumSegmentTree<int>((int)heavy_paths[i].size());
+	}
+
+	sort(cow_proficiency.begin(), cow_proficiency.end(), greater<pair<int, int>>());
+	for (auto &p : cow_proficiency) {
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:399]: consider using a structured binding

e.g. `for (auto &[thing_a, thing_b] : cow_proficiency)` or something

File: solutions/gold/usaco-696.mdx [hunk 15]
```diff
@@ -306,10 +307,122 @@ int main() {
...
+		if (heavy_index[p.second] != -1) {
+			ans[p.second] =
+			    segtrees[heavy_index[p.second]].range_sum(0, path_index[p.second] + 1);
+		} else ans[p.second] = cur_ans[p.second];
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:403]: braces

File: solutions/gold/usaco-696.mdx (hunk 6 above)
Comment (Vaishnav-K787) [reply solutions/gold/usaco-696.mdx:400]: it still throws a compilation error on usaco's website.

Comment (TheGamingMousse) [reply solutions/gold/usaco-696.mdx:400]: oh lol, then change it back

File: solutions/gold/usaco-696.mdx [hunk 16]
```diff
@@ -306,10 +307,124 @@ int main() {
...
+	if (heavy_child != -1) {
+		swap(heavy_paths[start], heavy_paths[heavy_child]);
+		heavy_paths[start].push_back(heavy_child);
+	}
+}
+
+int main() {
+	freopen("promote.in", "r", stdin);
+	freopen("promote.out", "w", stdout);
+	int n, current_value;
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:372]: why is this initialized here

File: solutions/gold/usaco-696.mdx [hunk 17]
```diff
@@ -306,10 +307,124 @@ int main() {
...
+	cin >> n;
+	vector<pair<int, int>> cow_proficiency(n);
+	for (int i = 0; i < n; ++i) {
+		cin >> cow_proficiency[i].first;
+		cow_proficiency[i].second = i;
+	}
+	vector edges(n, vector<int>()), heavy_paths(n, vector<int>());
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:379]: make these vector declarations on separate lines

File: solutions/gold/usaco-696.mdx [hunk 18]
```diff
@@ -306,10 +307,124 @@ int main() {
...
+
+	vector<int> sizes(n, 0), heavy_index(n, -1), path_index(n, -1), depth(n, 0),
```
Comment (TheGamingMousse) [inline solutions/gold/usaco-696.mdx:381]: same thing w/ declaring stuff on deparate lines

also, vectors autofill to 0

File: solutions/gold/usaco-696.mdx [hunk 19]
```diff
@@ -306,10 +307,124 @@ int main() {
...
+Sort the cows from largest to smallest proficiency. For each cow, perform a path update that adds one to all cows from itself to the president.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+// BeginCodeSnip{Segment Tree}
```
Comment (ryanchou-dev) [inline solutions/gold/usaco-696.mdx:323]: if this is copy-paste from the module, you should add (from the module) in the code snip title

File: solutions/gold/usaco-696.mdx (hunk 11 above)
Comment (bqi343) [reply solutions/gold/usaco-696.mdx:353]: or you can just put these variables into a class

**Review by TheGamingMousse — APPROVED** [review]
looks good to me

**Review by ryanchou-dev — COMMENTED** [review]
do you want to add some comments to the code?

otherwise, lgtm 👍

**Review by ryanchou-dev — APPROVED** [review]

**Pull request merged**